python/python/bystro/supervised_ppca/gf_dropout_pt.py

```diff
@@ -0,0 +1,299 @@
+            Covariates we wish to predict. For now lazy and assuming
+            logistic regression.
+
+        task : string,default='classification'
+            Is this prediction, multinomial regression, or classification
+
+        progress_bar : bool,default=True
+            Whether to print the progress bar to monitor time
+
+        seed : int,default=2021
+            The random number generator seed used to ensure reproducibility
+
+        Returns
+        -------
+        self : object
+            The model
+        """
+        self._test_inputs(X, y)
+        rng = np.random.default_rng(int(seed))
+        training_options = self.training_options
+        N, p = X.shape
+        self.p = p
+
+        W_, sigmal_, B_ = self._initialize_variables(X)
+        X, y = self._transform_training_data(X, 1.0*y)
+
+        if task == "classification":
+            sigm = nn.Sigmoid()
+            supervision_loss = nn.BCELoss()
+        elif task == "regression":
+            supervision_loss = nn.MSELoss()
+        else:
+            supervision_loss = nn.BCELoss()
+
+        trainable_variables = [W_, sigmal_, B_]
+
+        optimizer = torch.optim.SGD(
+            trainable_variables,
+            lr=training_options["learning_rate"],
+            momentum=training_options["momentum"],
+        )
+        eye = torch.tensor(np.eye(p))
+        one_s = torch.tensor(np.ones(self.n_supervised))
+        softplus = nn.Softplus()
+
+        _prior = self._create_prior()
+
+        myrange = trange if progress_bar else range
+
+        for i in myrange(training_options["n_iterations"]):
+            idx = rng.choice(
+                X.shape[0], size=training_options["batch_size"], replace=False
+            )
+            X_batch = X[idx]
+            y_batch = y[idx]
+
+            sigma = softplus(sigmal_)
+            WWT = torch.matmul(torch.transpose(W_, 0, 1), W_)
+            Sigma = WWT + sigma * eye
+
+            like_prior = _prior(trainable_variables)
+
+            # Generative likelihood
+            m = MultivariateNormal(torch.zeros(p), Sigma)
+            like_gen = torch.mean(m.log_prob(X_batch))
+
+            # Predictive lower bound
+            P_x, Cov = _get_projection_matrix(W_, sigma)
+            mean_z = torch.matmul(X_batch, torch.transpose(P_x, 0, 1))
+            eps = torch.rand_like(mean_z)
+            C1_2 = torch.cholesky(Cov)
+            z_samples = mean_z + torch.matmul(eps, C1_2)
+
+            y_hat = (
+                self.delta
+                * torch.matmul(z_samples[:, : self.n_supervised], one_s)
+                + B_
+            )
+            if task == "regression":
+                loss_y = supervision_loss(y_hat, y_batch)
+            else:
+                loss_y = supervision_loss(sigm(y_hat), y_batch)
+
+            WTW = torch.matmul(W_, torch.transpose(W_, 0, 1))
+            off_diag = WTW - torch.diag(torch.diag(WTW))
+            torch.norm(off_diag)
```

Comment:
what's going on here? This line appears to be a no-op, and torch.norm also seems to be deprecated: https://pytorch.org/docs/stable/generated/torch.norm.html

Reply:
Apparently "ruff --fix" isn't particularly reliable. I'll fix it and push again with matix_norm. But it's forcing independence between the different components, happens naturally in PPCA but not inherent with supervision loss.